sky/provision/kubernetes/network.py

```diff
@@ -57,39 +57,46 @@ def _open_ports_using_ingress(
     ports: List[int],
     provider_config: Dict[str, Any],
 ) -> None:
+    # Check if an ingress controller exists
     if not network_utils.ingress_controller_exists():
         raise Exception(
             'Ingress controller not found. '
             'Install Nginx ingress controller first: '
             'https://github.com/kubernetes/ingress-nginx/blob/main/docs/deploy/index.md.'  # pylint: disable=line-too-long
         )
 
-    for port in ports:
-        service_name = f'{cluster_name_on_cloud}-skypilot-service--{port}'
-        ingress_name = f'{cluster_name_on_cloud}-skypilot-ingress--{port}'
-        path_prefix = _PATH_PREFIX.format(
-            cluster_name_on_cloud=cluster_name_on_cloud, port=port)
+    # Prepare service names, ports,  for template rendering
+    service_details = [
+        (f'{cluster_name_on_cloud}-skypilot-service--{port}', port,
+         _PATH_PREFIX.format(cluster_name_on_cloud=cluster_name_on_cloud,
+                             port=port).rstrip('/').lstrip('/'))
+        for port in ports
+    ]
 
-        content = network_utils.fill_ingress_template(
-            namespace=provider_config.get('namespace', 'default'),
-            path_prefix=path_prefix,
-            service_name=service_name,
-            service_port=port,
-            ingress_name=ingress_name,
-            selector_key='skypilot-cluster',
-            selector_value=cluster_name_on_cloud,
-        )
+    # Generate ingress and services specs
+    content = network_utils.fill_ingress_template(
+        namespace=provider_config.get('namespace', 'default'),
+        service_details=service_details,
+        ingress_name=f'{cluster_name_on_cloud}-skypilot-ingress',
+        selector_key='skypilot-cluster',
+        selector_value=cluster_name_on_cloud,
+    )
```

Comment:
Could we add the PR description in the comments here or some other places in the code for why we batch the rules? : )
```suggestion
    content = network_utils.fill_ingress_template(
        namespace=provider_config.get('namespace', 'default'),
        service_details=service_details,
        ingress_name=f'{cluster_name_on_cloud}-skypilot-ingress',
        selector_key='skypilot-cluster',
        selector_value=cluster_name_on_cloud,
    )
```